src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c

```diff
@@ -858,26 +885,24 @@ PAL_SSLStreamStatus AndroidCryptoNative_SSLStreamWrite(SSLStream* sslStream, uin
     JNIEnv* env = GetJNIEnv();
     PAL_SSLStreamStatus ret = SSLStreamStatus_Error;
 
-    // int remaining = appOutBuffer.remaining();
-    // int arraySize = length > remaining ? remaining : length;
     // byte[] data = new byte[arraySize];
-    int32_t remaining = (*env)->CallIntMethod(env, sslStream->appOutBuffer, g_ByteBufferRemaining);
-    int32_t arraySize = length > remaining ? remaining : length;
-    jbyteArray data = make_java_byte_array(env, arraySize);
+    // data.setByteArrayRegion(0, length, buffer);
+    jbyteArray data = make_java_byte_array(env, length);
```

Comment:
Do we need to allocate a java array here or can we use NIO? `appOutBuffer` is a `ByteBuffer`. `ByteBuffer` has a `put` that accepts another `ByteBuffer`.

We can create a `ByteBuffer` over `buffer` with `jobject bufferByteBuffer = (*env)->NewDirectByteBuffer(env, buffer, length);`

Then we can `put` that in to appOutBuffer. That remove an allocation and a copy from a handshake.

Reply:
Just a thought, can be done as a follow up.

Reply:
Thanks for the suggestion, I'll give it a try.